### PR TITLE
*: add binary plan in stmt summary and slow log/query

### PR DESCRIPTION
**Pull request by time-and-fate**
### What problem does this PR solve?

Issue Number:  close https://github.com/pingcap/tidb/issues/35889

### What is changed and how it works?

- Add a field into `stmtCtx` to cache the binary plan.
- Add a field into the `SlowQueryLogItems` that contains the binary plan.
- Output the binary plan to the slow log in the format of `Binary_plan: tidb_decode_binary_plan('xxx')`
- Add a field into the `stmtSummaryByDigestElement` that contains the binary plan.
- Add a column into "statements_summary" and "slow_query" to output the binary plan.
- Add a global system variable to control if we should generate the binary plan for the execution of a statement.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: sessionctx/variable/session.go [hunk 1]
```diff
@@ -2318,6 +2322,10 @@ const (
 	SlowLogIsWriteCacheTable = "IsWriteCacheTable"
 )
 
+// GenerateBinaryPlan decides whether we should record binary plan in slow log and stmt summary.
+// It's controlled by the global variable `tidb_generate_binary_plan`.
+var GenerateBinaryPlan atomic2.Bool
```
Comment (AilinKid) [inline sessionctx/variable/session.go:2327]: it's a instance scope or cluster scope?

Comment (time-and-fate) [reply sessionctx/variable/session.go:2327]: It's cluster scope.
It's controlled by the global scope system variable. Once the system variable is updated, the `GenerateBinaryPlan` will be updated in all tidb.

**Review by AilinKid — APPROVED** [review]
LGTM

**Review by winoros — APPROVED** [review]

**Review by qw4990 — APPROVED** [review]

**Pull request merged**